doc/api/esm.md

```diff
@@ -313,50 +275,369 @@ If a package has no exports, setting `"exports": false` can be used instead of
+Aside from writing a stateless package (if JavaScript’s `Math` were a package,
+for example, it would be stateless as all of its methods are static), there are
+three ways to isolate state so that it’s shared between the potentially loaded
+CommonJS and ES module instances of the package:
+
+1. If possible, contain all state within an instantiated object. JavaScript’s
+   `Date`, for example, needs to be instantiated to contain state; if it were a
+   package, it would be used like this:
+
+    ```js
+    import date from 'date';
+    const someDate = new date();
+    // someDate contains state; date does not
+    ```
+
+   The `new` keyword isn’t required; a package’s function can return a new
+   object, or modify a passed-in object, to keep the state external to the
+   package.
+
+1. Isolate the state in one or more CommonJS files that are shared between the
+   CommonJS and ES module versions of the package. For example, if the CommonJS
+   and ES module entry points are `index.cjs` and `index.mjs`, respectively:
+
+    ```js
+    // ./node_modules/pkg/index.cjs
+    const state = require('./state.cjs');
+    module.exports.state = state;
+    ```
+
+    ```js
+    // ./node_modules/pkg/index.mjs
+    export state from './state.cjs';
+    ```
+
+   Even if `pkg` is used via both `require` and `import` in an application (for
+   example, via `import` in application code and via `require` by a dependency)
+   each reference of `pkg` will contain the same state; and modifying that
+   state from either module system will apply to both.
+
+   A package utilizing this pattern would not be usable unmodified in browsers
+   or in other environments that lack support for CommonJS.
+
+1. Write a package where state is stored globally. This is similar to the
+   previous approach, but instead of isolating state within a shared CommonJS
+   file it is attached to the global object, e.g.
+   `globalThis[Symbol.for('pkg@1.2.3:state')]`.
+
+    ```js
+    const stateSymbol = Symbol.for('pkg@1.2.3:state');
+    if (!(stateSymbol in globalThis)) {
+      Object.defineProperty(globalThis, stateSymbol, {value: {}});
+    }
+    const state = globalThis[singletonSymbol];
+    ```
+
+   This has the disadvantage of polluting the global namespace, but it is
+   compatible with non-CommonJS environments such as browsers. It also shares
+   state across multiple loaded numbered versions of the same package, e.g.
+   1.0.0 and 2.0.0, not just CommonJS vs ES module versions.
+
+For packages publishing separate CommonJS and ES module versions without a
+wrapper, in order for an `instanceof` comparison of the CommonJS and ES module
+versions of the package or of objects instantiated from each version to return
+`true`, [`Symbol.hasInstance`][] needs to be used to customize the behavior of
+`instanceof` for the exported object:
+
+```js
+const matcher = 'pkg@1.2.3:mainExport'; // Must be unique per package and export
+
+Object.defineProperties(exportedClassOrObject, {
```

Comment:
The MDN documentation for [Symbol.hasInstance](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/hasInstance) says to create `static [Symbol.hasInstance](instance)` so I this needs to be `exportedClass`.  Attaching to an instance does not work.

Reply:
So what about packages that aren’t classes?

Reply:
To clarify `function Pkg() {}` and `class Pkg() {}` are equal for this purpose, both are classes.  If you're asking something else then `instanceof` does not apply, the right hand side must be a class constructor.